include/libtorrent/packet_pool.hpp

```diff
@@ -0,0 +1,158 @@
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef TORRENT_PACKET_POOL_HPP
+#define TORRENT_PACKET_POOL_HPP
+
+#include "libtorrent/config.hpp"
+
+#include <boost/pool/pool.hpp>
+
+namespace libtorrent
+{
+
+	// used for out-of-order incoming packets
+	// as well as sent packets that are waiting to be ACKed
+	struct packet
+	{
+		// the last time this packet was sent
+		time_point send_time;
+
+		// the number of bytes actually allocated in 'buf'
+		std::uint16_t allocated;
+
+		// the size of the buffer 'buf' points to
+		std::uint16_t size;
+
+		// this is the offset to the payload inside the buffer
+		// this is also used as a cursor to describe where the
+		// next payload that hasn't been consumed yet starts
+		std::uint16_t header_size;
+
+		// the number of times this packet has been sent
+		std::uint8_t num_transmissions:6;
+
+		// true if we need to send this packet again. All
+		// outstanding packets are marked as needing to be
+		// resent on timeouts
+		bool need_resend:1;
+
+		// this is set to true for packets that were
+		// sent with the DF bit set (Don't Fragment)
+		bool mtu_probe:1;
+
+	#if TORRENT_USE_ASSERTS
+		int num_fast_resend;
+	#endif
+
+		// the actual packet buffer
+		std::uint8_t buf[1];
+	};
+
+	// single thread packet allocation packet pool
+	// can handle common cases of packet size by 3 pools
+	struct TORRENT_EXTRA_EXPORT packet_pool : single_threaded
+	{
+
+		packet *acquire(std::size_t allocate)
+		{
+			TORRENT_ASSERT(is_single_thread());
+			TORRENT_ASSERT(allocate <= std::numeric_limits<std::uint16_t>::max());
+
+			packet *p;
+
+			// every if block contains less or equal *_size check.
+			// "packet::allocate" field indicates only requested "packet::buf" size.
+			// (which may be less than claimed memory of corresponding slab)
+			if (allocate <= utp_header_size)
+			{
+				p = static_cast<packet*>(m_syn_slabs.malloc());
+			}
+			else if (allocate <= mtu_floor_size)
+			{
+				p = static_cast<packet*>(m_mtu_floor_slabs.malloc());
+			}
+			else if (allocate <= mtu_ceiling_size)
+			{
+				p = static_cast<packet*>(m_mtu_ceiling_slabs.malloc());
+			}
+			else
+			{
+				// fallback for huge packets, no need to use fixed size pool, just ordinary system malloc
+				p = static_cast<packet*>(malloc(sizeof(packet) + allocate));
+			}
+
+			p->allocated = static_cast<std::uint16_t>(allocate);
+			return p;
+		}
+
+		void release(packet *p)
+		{
+			TORRENT_ASSERT(is_single_thread());
+
+			if (p == nullptr) return;
+			if (p->allocated <= utp_header_size)
```

Comment:
and I think there should be a test that exercises this bug (along with a fix obviously)